src/main/java/org/truffleruby/core/FinalizationService.java

```diff
@@ -120,11 +126,34 @@ private void createFinalizationThread() {
             while (true) {
                 final FinalizerReference finalizerReference = (FinalizerReference) threadManager.runUntilResult(null,
                         () -> finalizerQueue.remove());
-                finalizerReference.getFinalizerActions().forEach(action -> action.run());
+                try {
+                    finalizerReference.getFinalizerActions().forEach(action -> action.run());
+                } catch (KillException e) {
+                    throw e;
+                } catch (RaiseException e) {
+                    handleFinalizeException(e);
+                } catch (Exception e) {
+                    // Do nothing, the finalizer thread must continue to process objects.
+                    if (context.getOptions().FINALIZE_PRINT_BACKTRACES) {
+                        e.printStackTrace();
+                    }
+                }
             }
         });
     }
 
+    private void handleFinalizeException(RaiseException exception) {
+        if (context.getOptions().FINALIZE_PRINT_BACKTRACES) {
```

Comment:
Does MRI give no output by default when an exception is raised in a finalizer?
It seems not, but `ruby -d` does show them. Could you check `-d` works for us too?